src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3811,8 +3778,56 @@ void os::large_page_init() {
     return;
   }
 
-  size_t large_page_size = Linux::setup_large_page_size();
-  UseLargePages          = Linux::setup_large_page_type(large_page_size);
+  // 2) Scan OS info
+  if (default_large_page_size == 0) {
+    // No large pages configured, return.
+    UseTransparentHugePages = false;
+    UseHugeTLBFS = false;
+    UseSHM = false;
+    return;
+  }
+  os::PageSizes all_large_pages = scan_multiple_page_support();
+
+  // 3) Consistency check and post-processing
+
+  // It is unclear if /sys/kernel/mm/hugepages/ and /proc/meminfo could disagree. Manually
+  // re-add the default page size to the list of page sizes to be sure.
+  all_large_pages.add(default_large_page_size);
+
+  for (size_t page_size = all_large_pages.largest(); page_size != 0;
+         page_size = all_large_pages.next_smaller(page_size)) {
+    _page_sizes.add(page_size);
+  }
+
+  LogTarget(Info, pagesize) lt;
+  if (lt.is_enabled()) {
+    LogStream ls(lt);
+    ls.print("Available page sizes: ");
+    _page_sizes.print_on(&ls);
+    ls.print("\n");
+    ls.print("Available large page sizes: ");
+    all_large_pages.print_on(&ls);
```

Comment:
Use `ls.cr()` instead of `print("\n")` and I'm also not sure there is any value to print the large page sizes separately as they are part of the first set.